A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,524 @@
+`override_host_status` value from its config's appropriate `DiscoveryMechanism`
+entry into the config for the `xds_override_host_experimental` policy. The
+diagram below shows how the configuration is passed down the hierarchy all the
+way from `cds_experimental` to `xds_override_host_experimental`.
+
+One of the existing policies (`xds_wrr_locality_experimental`,
+or `ring_hash_experimental`) is created as the child policy of
+`xds_override_host_experimental` as shown in the diagram below.
+
+![LB Policy Hierarchy Diagram](A55_graphics/lb-hierarchy.png)
+
+To propagate [`common_lb_config.override_host_status`][or-host-status] the
+following changes are required:
+
+* The [XdsClient][grpc-client-arch] will parse the
+  [common_lb_config.override_host_status][or-host-status] field from the
+  `Cluster` resource and copy the value to a new `override_host_status` field
+  in the CDS update that it sends to the watchers.
+
+* `cds_experimental` policy (which receives the CDS update) will copy the
+  value of `override_host_status` into the appropriate `DiscoveryMechanism`
+  entry's `override_host_status` field in the config of the
+  `xds_cluster_resolver_experimental` policy.
+
+* When the `xds_cluster_resolver_experimental` policy generates the config for
+  each priority, it will configure the `xds_override_host_experimental` policy
+  underneath the `xds_cluster_impl_experimental` policy. The generated config
+  for the `xds_override_host_experimental policy` will include the
+  `override_host_status` field from the `DiscoveryMechanism` associated with
+  that priority.
+
+### Notes
+
+Note that if stateful session affinity and retries (gRFC [A44][]) are both
+configured, all retry attempts will be routed to the same endpoint.
+
+### Temporary environment variable protection
+
+During initial development, this feature will be enabled by the
+`GRPC_EXPERIMENTAL_XDS_ENABLE_HOST_OVERRIDE` environment variable. This
+environment variable protection will be removed once the feature has
+proven stable. The environment variable protection will specifically apply
+to the following:
+
+* creation of `xds_override_host_experimental` LB policy
+
+* all of the logic related to processing of the cookie based stateful session
+  extension http_filter
+
+* processing of `common_lb_config.override_host_status` in the CDS update
+
+If the environment variable is unset or not true, none of the above will be
+enabled.
+
+## Rationale
+
+The stateful session affinity requirement is not satisfied by the
+[ring-hash LB policy][A42] because any time a backend host is removed or
+added, the hash mapping of 1/N client sessions gets affected (where N is the
+total number of backends) which would lead to dropped requests in some cases
+which is unacceptable.
+
+This feature avoids the problem by using HTTP cookies as described above.
+
+## Implementation
+
+The design will be implemented in gRPC C++, Java and Go.
```

Comment:
Please update this to reflect that we plan to implement it in C-core, but we currently have no plans to implement it in any other language.

Reply:
@markdroth I am not sure that's a correct statement. This support is necessary in C/C++, Python, Golang, and other languages we discussed offline.

Reply:
@va7ish this gRFC only states the current implementation plan based on the current priority of C/C++ that we were aware of when we started to work on the gRFC. But this statement in the gRFC does not preclude implementation in other languages once those languages are prioritized by the team and that we can continue discussing offline.